src/app/project/project.service.spec.ts

```diff
@@ -0,0 +1,180 @@
+            ]
+        });
+        this.service = TestBed.get(ProjectService);
+        this.http = TestBed.get(HttpTestingController); // Mapping Query Service requires http for plot subscriptions
+        this.backend = new MockBackend(this.http, MockConfig.MOCK_URL);
+        this.service.setProject(this.service.createDefaultProject());
+
+        clear_user_service_requests(this.http);
+
+        expect(this.http.match(req => true).length).toBe(0);
+
+        let user = TestBed.get(UserService);
+
+        let completed = null;
+        user.login({user: 'test', password: 'test_pw'})
+            .subscribe((login_response) => {
+                    expect(login_response).toBe(true);
+
+                    expect(user.getSession().user).toBe('test');
+                    expect(user.getSession().sessionToken).toBe('mockSessionToken')
+                },
+                (error) => {},
+                () => {
+                    completed = true;
+                }
+            );
+
+        this.backend.testLogin();
+        this.http.verify();
+        expect(user.getSession()).toEqual({
+            user: 'test',
+            sessionToken: 'mockSessionToken',
+            staySignedIn: true,
+            isExternallyConnected: false
+        });
+        expect(completed).toBeTruthy();
+    });
+
+    it('adds a plot', async () => {
+        let completed = false;
+        this.service.addPlot(Plot.fromDict({
+            name: 'test_plot',
+            operator: {
+                id: 0,
+                operatorType: {
+                    operatorType: 'r_script',
+                    resultType: 'plot',
+                    code: 'test_code'
+                } as RScriptTypeDict,
+                resultType: "plot",
+                projection: 'EPSG:3857',
+                attributes: [],
+                dataTypes: [],
+                units: [],
+                rasterSources: [],
+                pointSources: [],
+                lineSources: [],
+                polygonSources: [],
+            }
+        })).subscribe(
+            () => {},
+            (error) => {},
+            () => {
+                completed = true;
+            });
+        // TODO: Reverse engineer the URL.
+        setTimeout(() => {
+            const request = this.http.expectOne(MockConfig.MOCK_URL + '?time=&service=plot&request=&sessiontoken=mockSessionToken&crs=EPSG:3857&bbox=0,0,0,0&query=%7B%22type%22%3A%22r_script%22%2C%22params%22%3A%7B%22source%22%3A%22test_code%22%2C%22result%22%3A%22plot%22%2C%22plot_width%22%3A168%2C%22plot_height%22%3A168%7D%7D');
+            request.flush({type: 'png', data: 'dummy_data'});
+        });
+        this.http.verify();
+        await this.service.getProjectStream().subscribe((project) => {
+            expect(project.plots.length).toBe(1);
+        });
+        expect(completed).toBeTruthy();
+    });
+
+    it('clears plots', async () => {
+        this.service.clearPlots();
```

Comment:
Shouldn't there be at least one plot in the service to check for emptyness after calling `clearPlots`?

Reply:
The configureWaveTesting environment is a beforeEach environment internally. Since we add a plot in the configureWaveTesting we have one plot at every test atleast. That the plot is really added is secured by the 'adds plot' test. If that one fails the clear plots test becomes trivial, but atleast one test fails if something's wrong (in the other case we have one plot in the service).